nixos/modules/services/misc/mautrix-telegram.nix

```diff
@@ -0,0 +1,158 @@
+      cp '${overrideConfigFile}' "$out/config.yaml"
+      chmod +w "$out/config.yaml"
+
+      # generate the appservice registration token to be shared with the homeserver
+      ${pkgs.buildPackages.mautrix-telegram}/bin/mautrix-telegram \
+        --generate-registration \
+        --base-config='${pkgs.mautrix-telegram}/example-config.yaml' \
+        --config="$out/config.yaml" \
+        --registration="$out/registration.yaml"
+    '';
+  in mkIf cfg.enable {
+    # default minimal configuration
+    config.services.mautrix-telegram.config = {
+      appservice = rec {
+        database = mkDefault "sqlite:///${dataDir}/mautrix-telegram.db";
+        hostname = mkDefault "0.0.0.0";
+        port = mkDefault 8080;
+        address = mkDefault "http://localhost:${toString port}";
+      };
+
+      bridge = {
+        permissions."*" = mkDefault "relaybot";
+        relaybot.whitelist = mkDefault [ ];
+      };
+
+      # log to console/systemd instead of file
+      logging = {
+        version = 1;
+
+        formatters.precise.format = "[%(levelname)s@%(name)s] %(message)s";
+
+        handlers.console = {
+          class = "logging.StreamHandler";
+          formatter = "precise";
+        };
+
+        loggers = {
+          mau.level = mkDefault "INFO";
+          telethon.level = mkDefault "INFO";
+          aiohttp.level = mkDefault "INFO";
+        };
+
+        root = {
+          level = mkDefault "INFO";
+          handlers = [ "console" ];
+        };
+      };
+    };
+
+    systemd.services.mautrix-telegram = {
+      description = "Mautrix-Telegram, a Matrix-Telegram hybrid puppeting/relaybot bridge.";
+
+      wantedBy = [ "multi-user.target" ];
+      wants = [ "network-online.target" ] ++ optional (cfg.homeserverService != null) cfg.homeserverService;
+      after = [ "network-online.target" ] ++ optional (cfg.homeserverService != null) cfg.homeserverService;
+
+      preStart = ''
+        # run automatic database init and migration scripts
+        ${pkgs.mautrix-telegram.alembic}/bin/alembic -x config='${serviceConfig}/config.yaml' upgrade head
+      '';
+
+      serviceConfig = {
+        Type = "simple";
+        ExecStart = ''
+          ${pkgs.mautrix-telegram}/bin/mautrix-telegram \
+            --config='${serviceConfig}/config.yaml'
+        '';
+        Restart = "always";
+        DynamicUser = true;
+        WorkingDirectory = pkgs.mautrix-telegram; # required for the database migration scripts to be found
+        StateDirectory = baseNameOf dataDir;
+        # TODO: uncomment this once mautrix-telegram v0.5.3 is out (https://github.com/tulir/mautrix-telegram/pull/332)
+        #EnvironmentFile = cfg.environmentFile;
+      };
+    };
+
+    # TODO: define a common option to be used by all future Matrix homeserver implementations
+    # instead of assuming the use of Synapse.
+    services.matrix-synapse.app_service_config_files = [ "${serviceConfig}/registration.yaml" ];
```

Comment:
What if this bridge is running in a different machine than the homeserver? Can you please add an option to disable this ?

Reply:
Better yet, add an option `registerWithLocalHomeserver` (or something alike).

Reply:
Maybe should we drop the whole automatic registration.
It contains all-powerful tokens that can be used to do pretty much anything with the homeserver.
The module could require the user to provide a registration file explicitly.

Reply:
Automatic registration removed.
This allows the user to explicitly choose which homeserver implementation to use, possibly running on a remote machine.